Add new Trove translator

wragge opened the pull request:
The 'National Library of Australia - Australian Newspapers' translator was broken by a recent update to the site. I took the opportunity to create a new translator for (almost) the whole of Trove, of which the newspapers are now just one part. If this translator is ok, then the old 'Australian Newspapers' translator can be removed from the repository.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+*/
+
+function detectWeb(doc, url) {
+	if (url.match(/\/result\?/) || url.match(/\/newspaper\/page/)) {
```
here and below, we prefer using .indexOf("string") != -1 to test for literal strings and .search(/regex/) != -1 where you need rege capabilities. It's a lot more efficient.

wragge replied:
Done.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+		// First you need to tell the service to start generating the PDF.
+		// You get back a hash id to include in the PDF url.
+
+		var renditionURL = "http://trove.nla.gov.au/newspaper/rendition/"
```
using a relative link should be more stable, e.g. if the site supports https (i.e. `"/newspaper/rendition"`)

wragge replied:
Done.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+
+			var timeout = Date.now() + 5000;
+			while (Date.now() < timeout){}
+			item.attachments.push({url: renditionURL + articleID + '.3.pdf?followup=' + hashID, title: item.publicationTitle + ', ' + item.date + ', p. ' + item.pages, mimeType:'application/pdf'});
```
we perfer a generic label like `Trove Full Text PDF` for the title here (the filename is generated automatically by Zotero)

wragge replied:
Done -- named 'Trove Newspaper PDF' (not strictly full text as it's actually the images).

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+	var thumbnailURL;
+
+	// Remove all params from url
+	var workURL = url.match(/http.*\/work\/\d+/)[0];
```
we're positive this always matches? Otherwise the [0] throws an error here which should be avoided.

wragge replied:
Improved -- uses indexOf and slice.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+				"title": "THIRROUL - Hotels - Rex Hotel",
+				"creators": [
+					{
+						"firstName": "1910-1981, William A. (William Alan)",
```
we should be able to do better than this, no? Can we clean up names some?

wragge replied:
I think it's fairly rare for the dates to creep in to the BibTex, but I've added a function to clean dates out of first names.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,326 @@
+				"date": "1980",
+				"ISBN": "9780908065073",
+				"abstractNote": "In 14 libraries. 24 p. : ill. ; 22 cm. Wragge, Clement L. (Clement Lindley), 1852-1922. South Australia. Climate, 1883-1884. Meteorologists -- South Australia -- Biography. South Australia -- Climate -- History.",
+				"extra": "Reprinted from Good words for 1887/ edited by Donald Macleod, published: London: Isbister and Co",
```
which field in Bibtex is this? I'd generally prefer this in a note, but that might require fixing bibtex import rather than here.

wragge replied:
It's in the 'note' field. I haven't done anything about this.

adam3smith replied:
could you push this into a note and remove from extra? I'd rather not change bibtex import in general, but this is a poor fit for extra.

adam3smith commented:
See comments inline; looks good to me otherwise.

wragge commented:
Thanks for the comments. I think I've addressed them all.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,340 @@
+function getSearchResults(doc, url) {
+	var items = {};
+	var results;
+	if (url.match(/\/result\?/)) {
```
some more `match` to be replaced here

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,340 @@
+	var results;
+	if (url.match(/\/result\?/)) {
+		results = ZU.xpath(doc, "//div[@id='mainresults']//li/dl/dt/a");
+	} else if (url.match(/\/newspaper\/page/)) {
```
here

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,340 @@
+}
+
+function scrape(doc, url) {
+	if (url.match(/\/newspaper\/article\//i)) {
```
and here

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,340 @@
+			delete item.itemID;
+			delete item.type;
+			if (thumbnailURL !== null) {
+				item.attachments.push({url: thumbnailURL, title: item.title, mimeType:'image/jpeg'});
```
and if you could use a generic title here, too, as above. I overlooked this on the first go.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,340 @@
+			// So I took out the ping call completely and just put in a 5 second pause which seems to work ok.
+			// Note that the service itself uses a 5 second delay if the ping fails.
+
+			var timeout = Date.now() + 5000;
```
@aurimasv @zuphilip  -- that's the only potentially problematic bit I see otherwise. Looks good to me, any concerns from your side?

wragge replied:
As I note in the comment, this is the only way I can see to reliably get the PDF. If I could get access to the response object I could do a quick check, but if that came back "Locked" I'd still need to do the timeout. This is exactly what the Trove UI itself does -- pings, if status is locked it includes a 5 second timeout (except that it can use setTimeout which seems to be blocked within Zotero).

zuphilip replied:
Yes, this looks problematic and it could lead to larger waiting times.
E.g. the 20 items of [this search](http://trove.nla.gov.au/book/result?l-australian=y&q=wragge) would lead to 100 sec waiting time, which is more than 1,5 minutes!

adam3smith replied:
but is this a trade-off worth making? Would it be that confusing if a download of 20 items with images takes a while? 
The way I understand, the alternative is removing that component altogether, right?
One issue I'm seeing is that this would halt the entire import process. I haven't thought this through, but could we put the attachment process into an asynchronous function, so the item saves and we're "just" waiting for the attachment?

wragge replied:
Yeah but you wouldn't be waiting 100 sec without anything happening. You'd see an item was being processed every 5 sec or so. And in any case if you asked any user would they rather wait for 100 sec, or manually view, download, and add to Zotero every individual item I can tell you what the answer would be...

The automatic download of newspaper article PDFs from Trove has brought 'wows' in various workshop/demos I've given -- it's been a very popular feature and I would not want to have it removed.

zuphilip replied:
Okay, I am convinced that the waiting time is worth the advantages.

(OT @wragge I like your presentation of your digitalized newspapers very much and maybe I can contact you on that separately.)

adam3smith commented:
@wragge  -- you may want to wait with fixing my other nits above until we hear back from a 2nd person about the timeout, up to you.

wragge commented:
Argh sorry didn't mean for that last commit to be added to this pull request. It's only very minor though...

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		results = ZU.xpath(doc, "//ol[@class='list-unstyled articles']/li/h4/a");
+	}
+	for (var i=0; i<results.length; i++) {
+		var url = results[i].href;
```
The variable `url` is a parameter of the function and therefore the variable is already defined. I would prefer here to use another name for this new variable.

adam3smith replied:
ah yes, most definitely, good catch.

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		// First you need to tell the service to start generating the PDF.
+		// You get back a hash id to include in the PDF url.
+
+		var renditionURL = "/newspaper/rendition/"
```
Missing semicolon

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		// You get back a hash id to include in the PDF url.
+
+		var renditionURL = "/newspaper/rendition/"
+		var prepURL = renditionURL + articleID + '/level/3/prep'
```
Missing semicolon

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	"Map": "map",
+	"Unpublished": "manuscript",
+	"Published": "document"
+}
```
Missing semicolon

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	for (var i=0; i<results.length; i++) {
+		var url = results[i].href;
+		var title = ZU.trimInternal(results[i].textContent);
+		items[url] = title;
```
and then here, too, of course.

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	newCreators = [];
+	for (var i = 0; i < creators.length; i++) {
+		var creator = creators[i];
+		creator.firstName = creator.firstName.replace(/\(?\d{4}-\d{0,4}\)?,?/, "").trim()
```
Missing semicolon

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		creator.firstName = creator.firstName.replace(/\(?\d{4}-\d{0,4}\)?,?/, "").trim()
+		newCreators.push(creator);
+	}
+	return newCreators
```
Missing semicolon

zuphilip commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+			ZU.HTTP.doGet(textURL, function(text) {
+				item.notes.push({note: text.trim()});
+				item.complete();
+			});
```
It looks that the `hashID` is not needed here, and therefore I guess you can put this lines also outside the outer `doGet`.

wragge replied:
Where would the item.complete() go then?

zuphilip replied:
Ah, I missed that one. Maybe write a comment then.

aurimasv commented on National Archives of Australia.js:
```diff
@@ -207,14 +207,10 @@ function scrapeSeries(doc, url) {
-		mimeType: 'text/html',
-		snapshot: false
-	})
+	item.url = 'http://www.naa.gov.au/cgi-bin/Search?O=S&Number=' + seriesNumber;
```
Update tests please.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+
+function detectWeb(doc, url) {
+	if (url.indexOf('/result?') != -1 || url.indexOf('/newspaper/page') != -1) {
+		return "multiple";
```
make sure there are actual results to import (i.e. call `getSearchResults`)

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	} else if (url.indexOf('/newspaper/article') != -1) {
+		return "newspaperArticle";
+	} else if (url.indexOf('/work/') != -1) {
+		return "document";
```
Since this is not actually the item type that will be imported, the test that you included at the bottom of the translator will fail. I would probably use the most likely item type here (seems to be artwork).

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		results = ZU.xpath(doc, "//div[@id='mainresults']//li/dl/dt/a");
+	} else if (url.indexOf('/newspaper/page') != -1) {
+		results = ZU.xpath(doc, "//ol[@class='list-unstyled articles']/li/h4/a");
+	}
```
And if neither?

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		var title = ZU.trimInternal(results[i].textContent);
+		items[url] = title;
+	}
+	return items;
```
generally, our pattern has been to return `false` if there were no results and the object otherwise. It's a bit annoying to check that there is nothing in the object otherwise.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+			// Note that the service itself uses a 5 second delay if the ping fails.
+
+			var timeout = Date.now() + 5000;
+			while (Date.now() < timeout){}
```
This is not good. The way we process translator code may have changed recently (in prep for e10s, perhaps), but this would effectively lock up the browser for 5 seconds, no? I know we don't expose a number of global (window) functions in translators, though you may still be able to get to them via document object (`document.defaultView`). See if you can use that to get `setTimeout`. If no, and this does lock up the browser, then I would say this is not acceptable.

aurimasv replied:
Actually, this may not be possible. Since setTimeout is async, you would need to call incrementAsyncProcesses before and decrementAsyncProcesses at the end of the call. Otherwise Zotero doesn't know that it needs to wait for something else to finish. Those functions are not exposed to translators. So, if the while loop locks up the browser, which I think it does, this would not be acceptable.

wragge replied:
For reasons explained above the PDF attachment needs to be there. Users will cope with a 5 sec delay.

adam3smith replied:
@aurimasv is raising a different issue though: A 5 sec *delay* is one thing, and I'm fine with that. A 5 sec (let alone 1min +) *browser freeze* I don't think is acceptable behavior for Zotero to cause, no matter how useful the PDF attachment. 
I haven't actually tested this yet, so if the browser doesn't freeze, I completely agree it's worth it.

wragge replied:
Not sure what is meant by freeze. If I select multiple newspaper articles using Zotero I can switch tabs while I'm waiting and do whatever else I want in the browser -- so it's not frozen?

dstillman replied:
@wragge Are you testing this with Standalone or Firefox? I'm pretty sure this would lock up Firefox. With Standalone, it might run in the connector, but I think it would lock up the tab and peg the CPU in its helper process, even if you can still switch to another tab.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+
+			var timeout = Date.now() + 5000;
+			while (Date.now() < timeout){}
+			item.attachments.push({url: renditionURL + articleID + '.3.pdf?followup=' + hashID, title: 'Trove newspaper PDF', mimeType:'application/pdf'});
```
use multiple lines

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+			while (Date.now() < timeout){}
+			item.attachments.push({url: renditionURL + articleID + '.3.pdf?followup=' + hashID, title: 'Trove newspaper PDF', mimeType:'application/pdf'});
+
+			// Get the OCRd text and save in a note.
```
I'm not sure that we should be getting the OCRd text and attaching the PDF. I think one or the other, but not both.

wragge replied:
Yes both. PDF contains the article images. OCR is searchable text.

zuphilip replied:
Doesn't have the pdf the text as well? Zotero can index pdfs and search them as well.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	"Printed music": "book",
+	"Map": "map",
+	"Unpublished": "manuscript",
+	"Published": "document"
```
not a great choice, since it has very limited fields and is not cited well. We generally default to journalArticle or book.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+}
+	
+
+function checkType(string) {
```
Could you add a comment describing what this function does? What's the expected format of the input to this function?

Also, please use something more descriptive than "string"

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	
+
+function checkType(string) {
+	types = string.split("; ");
```
missing `var ` (here and below)

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+function checkType(string) {
+	types = string.split("; ");
+	newString = types.join(" ");
+	if (newString in troveTypes) {
```
`.hasOwnProperty` is actually what you want

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		return troveTypes[newString];
+	} else {
+		while (types.length > 0) {
+			types.pop();
```
pop at the end of the while and skip the whole if statement above

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+}
+
+function cleanCreators(creators) {
+	newCreators = [];
```
`var`

adam3smith replied:
this still needs to get fixed: `var newCreators = [];`

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	for (var i = 0; i < creators.length; i++) {
+		var creator = creators[i];
+		creator.firstName = creator.firstName.replace(/\(?\d{4}-\d{0,4}\)?,?/, "").trim()
+		newCreators.push(creator);
```
You're modifying the passed-in `creators` object anyway. Why return a new object in the end?

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+}
+
+function scrapeWork(doc, url) {
+	var thumbnailURL;
```
nit: move closer to where it's used

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+	var thumbnailURL;
+
+	// Remove all params from url
+	var workURL = url.slice(0, url.indexOf('?'));
```
I prefer `.replace(/[?#].*/, '')`, since the logic above does not account for `?` missing altogether. (`var a = 'abcd'; a.slice(0, a.indexOf('e')) == 'abc'`)

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+		thumbnailURL = ZU.xpathText(doc, "//a/img[@class='mosaic ui-shdw']/@src");
+	} else {
+		// It's a work -- so thumbnails are different
+		thumbnailURL = ZU.xpathText(doc, "//li[@class='imgfirst']//img/@src");
```
What happens here with the bibtexURL? Is it ok for it not to have the ID? Please comment

adam3smith replied:
still need an answer for this.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+			// This gives a better version-aware url.
+			item.url = ZU.xpathText(doc, "//meta[@property='og:url']/@content");
+			item.abstractNote = ZU.xpathText(doc, "//meta[@property='og:description']/@content");
+			delete item.itemID;
```
generally speaking, you don't need to worry about deleting these. If they don't belong to a type, they will be omitted automatically.

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+			delete item.itemID;
+			delete item.type;
+			if (thumbnailURL !== null) {
+				item.attachments.push({url: thumbnailURL, title: 'Trove thumbnail image', mimeType:'image/jpeg'});
```
use multiple lines, but is this actually valuable? Could we not attach the full image? (I'm guessing this may be too complicated) Attaching the thumbnail may give people the wrong impression that they have a full copy of the image (since we don't attach thumbnails or incomplete articles anywhere else). It would probably be more useful to attach a link to the website hosting the full image.

wragge replied:
Trove is an aggregator -- images come from hundreds of different sources. Thumbnail links are the only reliable ones. I believe users will find them valuable as a quick check on the identity of the image. But I will also attach a link to the providing organisations where they are provided (this will be a link to an HTML page not to an image).

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+				"title": "THIRROUL - Hotels - Rex Hotel",
+				"creators": [
+					{
+						"firstName": "William A. (William Alan)",
```
Could we clean this up?

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+					}
+				],
+				"date": "1980",
+				"ISBN": "9780908065073",
```
Was this imported with Scaffold? I thought Scaffold would hyphenate these automatically, no?

aurimasv commented on Trove.js:
```diff
@@ -0,0 +1,337 @@
+				"tags": [],
+				"notes": [
+					{
+						"note": "'WRAGGE' - we have received a copy of the above, which is a journal devoted chiefly to the science of meteorology. It is owned and conducted by Mr. Clement Wragge."
```
Is this a useful note? Are notes in general from this website useful?

zuphilip replied:
It is the full text of the short newspaper article, see http://trove.nla.gov.au/newspaper/article/70068753 (the OCR version of the article is saved into a note)

wragge replied:
Yes it's useful

aurimasv commented:
Please take a look at some of the comments that got hidden by your recent commits. Looks like we overlapped.

wragge commented:
I think I've addressed most of the issues raised.

To get around the PDF timeout issue, I've created a proxy server which returns the PDF url when it's ready.

While I was thinking about this, they changed the location of the BibTex files for newspaper articles, so I had to fix that as well.

adam3smith commented:
just wanted to say that this isn't being ignored; it just hit at a very busy time, I'll review asap, but worst case might take another two weeks. Apologies.

wragge commented:
Ok, no worries. Thanks for letting me know.

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,334 @@
+	},
+	{
+		"type": "web",
+		"url": "http://trove.nla.gov.au/newspaper/result?l-australian=y&q=wragge",
```
I get an error testing on this page. The Trove side throws
`main / UI Layout Initialization Warning
The layout-container "main layout / DIV#ui-layout-main" has no height.
Therefore the layout is 0-height and hence 'invisible'!`

adam3smith commented on Trove.js:
```diff
@@ -0,0 +1,334 @@
+	},
+	{
+		"type": "web",
+		"url": "http://trove.nla.gov.au/book/result?l-australian=y&q=wragge",
```
this one works OK .

zuphilip commented:
Superseeded by #1208.
